wdl-grammar/src/tree/dive.rs

```diff
@@ -0,0 +1,259 @@
+///
+/// There are also facilities to ignore certain subtrees within the tree via the
+/// `ignore_predicate` callback. If the entire tree is desired, one can simply
+/// have the `ignore_predicate` always return `false`.
+#[derive(Debug)]
+pub struct DiveIterator<L, M, I>
+where
+    L: Language,
+    M: Fn(&rowan::SyntaxElement<L>) -> bool,
+    I: Fn(&rowan::SyntaxNode<L>) -> bool,
+{
+    /// The queue of nodes to be visited.
+    nodes: VecDeque<rowan::SyntaxElement<L>>,
+    /// The function that evaluates when checking if an element matches.
+    match_predicate: M,
+    /// The function that evaluates when checking if the children underneath a
+    /// node in the tree should be ignored.
+    ignore_predicate: I,
+    /// A pointer to the [`Language`].
+    _phantom: PhantomData<L>,
+}
+
+impl<L, M, I> DiveIterator<L, M, I>
+where
+    L: Language,
+    M: Fn(&rowan::SyntaxElement<L>) -> bool,
+    I: Fn(&rowan::SyntaxNode<L>) -> bool,
+{
+    /// Creates a new [`DiveIterator`].
+    ///
+    /// Note that this iterator traverses elements as a breadth-first traversal
+    /// (otherwise known as level-order).
+    pub fn new(root: rowan::SyntaxElement<L>, match_predicate: M, ignore_predicate: I) -> Self {
+        let mut nodes = VecDeque::new();
+        nodes.push_back(root);
+
+        Self {
+            nodes,
+            match_predicate,
+            ignore_predicate,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<L, M, I> Iterator for DiveIterator<L, M, I>
+where
+    L: Language,
+    M: Fn(&rowan::SyntaxElement<L>) -> bool,
+    I: Fn(&rowan::SyntaxNode<L>) -> bool,
+{
+    type Item = rowan::SyntaxElement<L>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(element) = self.nodes.pop_front() {
+            match &element {
+                rowan::SyntaxElement::Node(node) => {
+                    if !(self.ignore_predicate)(node) {
+                        self.nodes.extend(node.children_with_tokens());
+                    }
+                }
+                // NOTE: tokens have no children to explore.
+                rowan::SyntaxElement::Token(_) => {}
+            }
+
+            if (self.match_predicate)(&element) {
+                // NOTE: this is an inexpensive clone of a red node.
+                return Some(element.clone());
+            }
+        }
+
+        None
+    }
+}
+
+impl<L, M, I> FusedIterator for DiveIterator<L, M, I>
+where
+    L: Language,
+    M: Fn(&rowan::SyntaxElement<L>) -> bool,
+    I: Fn(&rowan::SyntaxNode<L>) -> bool,
+{
+}
+
+/// Elements of a syntax tree upon which a dive can be performed.
+pub trait Divable<L>
+where
+    L: Language,
+{
+    /// Visits every element in a syntax tree and accumulates all
+    /// [`SyntaxElement`](rowan::SyntaxElement)s for which the `match_predicate`
+    /// evaluates to `true`.
+    ///
+    /// No guarantees are made about the order in which elements will be
+    /// traversed or collected.
```

Comment:
As there's no accumulation/collection taking place, I think we could simply say the following:

```suggestion
    /// Iterates over every element in the tree at the current root and yields the
    /// elements for which the given `match_predicate` evaluates to `true`.
    ///
    /// No guarantees are made about the order in which elements will be traversed.
```

Reply:
I do wonder if these methods should just return `DiveIterator` so we can make a guarantee about the ordering rather than `impl Iterator` (perhaps even renaming `DiveIterator` to `BreadthFirstIterator` to make it even more explicit).

I'm sure you have a reason, though, for wanting the ordering to be unspecified in the `Divable` trait.

Reply:
@peterhuene and I discussed offline—we decided to wrap the provided `SyntaxNode::preorder_with_tokens` iterator.